File: appAndroid/common-compose/src/main/java/com/gchristov/newsfeed/commoncompose/elements/AppButton.kt
package com.gchristov.newsfeed.commoncompose.elements

import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.padding
import androidx.compose.material.ButtonDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.unit.dp
import com.gchristov.newsfeed.commoncompose.theme.Theme
import com.gchristov.newsfeed.commondesign.R

@Composable
fun AppButton(
    modifier: Modifier = Modifier,
    onClick: () -> Unit,
    text: String
) {
    androidx.compose.material.Button(
        modifier = modifier,
        onClick = onClick,
        colors = ButtonDefaults.buttonColors(
            backgroundColor = Theme.contentColors.action
        )
    ) {
        AppText(
            text = text,
            color = colorResource(R.color.white)
        )
    }
}

@Composable
fun AppIconButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    icon: ImageVector,
    tint: Color = Theme.contentColors.primary,
    contentDescription: String? = null,
) {
    androidx.compose.material.IconButton(
        onClick = onClick,
        modifier = modifier,
        interactionSource = remember { MutableInteractionSource() },
    ) {
        AppIcon(
            imageVector = icon,
            tint = tint,
            contentDescription = contentDescription
        )
    }
}

@Composable
fun AppTextButton(
    modifier: Modifier = Modifier,
    onClick: () -> Unit,
    text: String
) {
    AppText(
        modifier = modifier
            .clickable { onClick() }
            .padding(vertical = 8.dp),
        text = text,
        color = Theme.contentColors.action,
    )
}

File: ios/common-swiftui/Sources/Elements/AppButton.swift
import SwiftUI

public struct AppButton: View {
    @EnvironmentObject var theme: Theme
    private let text: String
    private let onClick: () -> ()
    
    public init(
        text: String,
        onClick: @escaping () -> ()
    ) {
        self.text = text
        self.onClick = onClick
    }
    
    public var body: some View {
        Button {
            onClick()
        } label: {
            AppText(
                text: text,
                color: Color.appWhite
            )
        }
        .padding(.top, 8)
        .padding(.bottom, 8)
        .padding(.leading, 16)
        .padding(.trailing, 16)
        .background(theme.contentColors.action)
        .cornerRadius(3)
        .shadow(radius: 2, x: 0.2, y: 0.5)
    }
}

public struct AppIconButton: View {
    @EnvironmentObject var theme: Theme
    private let imageName: String
    private let tint: Color?
    private let contentDescription: String
    private let onClick: () -> ()
    
    public init(
        imageName: String,
        tint: Color? = nil,
        contentDescription: String,
        onClick: @escaping () -> ()
    ) {
        self.imageName = imageName
        self.tint = tint
        self.contentDescription = contentDescription
        self.onClick = onClick
    }
    
    public var body: some View {
        Button {
            onClick()
        } label: {
            AppIcon(
                imageName: imageName,
                tint: tint
            )
        }
        .accessibilityLabel(contentDescription)
    }
}

public struct AppTextButton: View {
    @EnvironmentObject var theme: Theme
    private let text: String
    private let onClick: () -> ()
    
    public init(
        text: String,
        onClick: @escaping () -> ()
    ) {
        self.text = text
        self.onClick = onClick
    }
    
    public var body: some View {
        Button {
            onClick()
        } label: {
            AppText(
                text: text,
                color: theme.contentColors.action
            )
        }
        .padding(.top, 8)
        .padding(.bottom, 8)
    }
}
